hw/opentitan/ot_otp_engine.c

```diff
@@ -0,0 +1,2969 @@
+    bool is_secret = ot_otp_engine_is_secret(s, part_ix);
+    bool is_digest = ot_otp_engine_is_part_digest_offset(s, part_ix, address);
+    bool is_zer = ot_otp_engine_is_part_zer_offset(s, part_ix, address);
+
+    /* "in all partitions, the digest itself is ALWAYS readable." */
+    if (!is_digest && !is_zer && !is_readable) {
+        qemu_log_mask(LOG_GUEST_ERROR,
+                      "%s: %s: partition %s @ 0x%04x not readable\n", __func__,
+                      s->ot_id, ot_otp_engine_part_name(s, part_ix), address);
+        ot_otp_engine_dai_set_error(s, OT_OTP_ACCESS_ERROR);
+        return;
+    }
+
+    unsigned part_offset = address - ot_otp_engine_part_data_offset(s, part_ix);
+    unsigned part_waddr = part_offset >> 2u;
+    bool do_ecc =
+        s->part_descs[part_ix].integrity && ot_otp_engine_is_ecc_enabled(s);
+
+    DAI_CHANGE_STATE(s, OT_OTP_DAI_READ_WAIT);
+
+    uint32_t data_lo, data_hi;
+    unsigned err = 0;
+    unsigned cell_count = sizeof(uint32_t) + (do_ecc ? sizeof(uint16_t) : 0);
+
+    const uint32_t *data = (const uint32_t *)s->otp->data;
+    /* parenthesis inform the C linter sizeof() call is valid with 'data */
+    data += (ot_otp_engine_part_data_offset(s, part_ix) / sizeof(uint32_t));
+
+    if (is_wide || is_digest) {
+        /* 64-bit requests */
+        part_waddr &= ~0b1u;
+
+        g_assert((part_waddr + 1u) * sizeof(uint32_t) <
+                 s->part_descs[part_ix].size);
+
+        data_lo = data[part_waddr];
+        data_hi = data[part_waddr + 1u];
+
+        if (do_ecc) {
+            unsigned ewaddr = address >> 3u;
+            g_assert(ewaddr < s->otp->ecc_size);
+            uint32_t ecc = s->otp->ecc[ewaddr];
+            if (ot_otp_engine_is_ecc_enabled(s)) {
+                data_lo =
+                    ot_otp_engine_verify_ecc(s, data_lo, ecc & 0xffffu, &err);
+                data_hi =
+                    ot_otp_engine_verify_ecc(s, data_hi, ecc >> 16u, &err);
+            }
+        }
+
+        cell_count *= 2u;
+    } else {
+        /* 32-bit request */
+        g_assert(part_waddr * sizeof(uint32_t) < s->part_descs[part_ix].size);
+
+        data_lo = data[part_waddr];
+        data_hi = 0u;
+
+        if (do_ecc) {
+            unsigned ewaddr = address >> 3u;
+            g_assert(ewaddr < s->otp->ecc_size);
+            uint32_t ecc = s->otp->ecc[ewaddr];
+            if ((address >> 2u) & 1u) {
+                ecc >>= 16u;
+            }
+            if (ot_otp_engine_is_ecc_enabled(s)) {
+                data_lo =
+                    ot_otp_engine_verify_ecc(s, data_lo, ecc & 0xffffu, &err);
+            }
+            cell_count = 4u + 2u;
+        } else {
+            cell_count = 4u;
+        }
+    }
+
+    if (is_secret && !is_digest) {
+        /*
+         * if the partition is a secret partition, OTP storage is scrambled
+         * except the digest and the zeroification fields
+         */
```

Comment:
I suspect this got lost from the DJ OTP in the backport?
```suggestion
    if (is_secret && !(is_zer || is_digest)) {
        /*
         * if the partition is a secret partition, OTP storage is scrambled
         * except the digest and the zeroification fields
         */
```

Reply:
You're right